File: src/viem/storageLens.ts
import {
  Address,
  Hex,
  PublicClient,
  RpcTransactionRequest,
  decodeFunctionResult,
  encodeFunctionData,
  toHex,
} from "viem";
import { EphemeralStorageLens__factory } from "../../typechain";

type StateOverrides = {
  [address: Address]: {
    balance?: Hex;
    nonce?: Hex;
    code?: Hex;
    stateDiff?: {
      [slot: Hex]: Hex;
    };
  };
};

/**
 * Call a contract with the given state overrides.
 * @param tx The transaction request.
 * @param overrides The state overrides.
 * @param publicClient A JSON RPC provider that supports `eth_call` with state overrides.
 * @param blockNumber Optional block number to use for the call.
 */
export async function staticCallWithOverrides(
  tx: RpcTransactionRequest,
  overrides: StateOverrides,
  publicClient: PublicClient,
  blockNumber?: bigint,
): Promise<Hex> {
  return (await publicClient.request({
    method: "eth_call",
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    params: [tx, blockNumber ? toHex(blockNumber) : "latest", overrides],
  })) as Hex;
}

/**
 * Batch `eth_getStorageAt` RPC calls in a single `eth_call` by overriding the target contract's
 * deployed bytecode with `EphemeralStorageLens`
 * @param address The contract address to fetch storage from.
 * @param slots The storage slots to fetch.
 * @param publicClient A JSON RPC provider that supports `eth_call` with state overrides.
 * @param blockNumber Optional block number to query.
 */
export async function getStorageAt(
  address: Address,
  slots: Hex[],
  publicClient: PublicClient,
  blockNumber?: bigint,
): Promise<readonly Hex[]> {
  const overrides = {
    [address]: {
      code: "0x60806040526004361061001e5760003560e01c8063dbd035ff14610023575b600080fd5b61003661003136600461007c565b61004c565b60405161004391906100f3565b60405180910390f35b60606020600052816020528160051b6040016040845b803554825260209182019101828210610062575050806000f35b6000806020838503121561008f57600080fd5b823567ffffffffffffffff8111156100a657600080fd5b8301601f810185136100b757600080fd5b803567ffffffffffffffff8111156100ce57600080fd5b8560208260051b84010111156100e357600080fd5b6020919091019590945092505050565b602080825282518282018190526000918401906040840190835b8181101561012b57835183526020938401939092019160010161010d565b50909594505050505056fea164736f6c634300081a000a" as Hex,
    },
  };
  const data = await staticCallWithOverrides(
    {
      to: address,
      from: "0x0000000000000000000000000000000000000000",
      data: encodeFunctionData({
        abi: EphemeralStorageLens__factory.abi,
        functionName: "extsload",
        args: [slots],
      }),
    },
    overrides,
    publicClient,
    blockNumber,
  );
  return decodeFunctionResult({
    abi: EphemeralStorageLens__factory.abi,
    functionName: "extsload",
    data,
  });
}

File: src/storage_lens.rs
//! ## Storage Lens
//!
//! The storage lens module provides a function to batch `eth_getStorageAt` RPC calls in a single
//! `eth_call` by overriding the target contract's deployed bytecode with `EphemeralStorageLens`.

use crate::{bindings::ephemeralstoragelens::EphemeralStorageLens, error::Error};
use alloc::vec::Vec;
use alloy::{
    eips::BlockId,
    network::Network,
    primitives::{Address, B256},
    providers::Provider,
    rpc::types::state::{AccountOverride, StateOverride},
};

/// Batch `eth_getStorageAt` RPC calls in a single `eth_call` by overriding the target contract's
/// deployed bytecode with `EphemeralStorageLens`
///
/// # Arguments
///
/// * `address`: The contract address to fetch storage from
/// * `slots`: The storage slots to query
/// * `provider`: The alloy provider
/// * `block_id`: Optional block id to query
///
/// ## Returns
///
/// The storage values at the given slots
#[inline]
pub async fn get_storage_at<N, P>(
    address: Address,
    slots: Vec<B256>,
    provider: P,
    block_id: Option<BlockId>,
) -> Result<Vec<B256>, Error>
where
    N: Network,
    P: Provider<N>,
{
    // override the deployed bytecode at `address`
    let state = StateOverride::from_iter([(
        address,
        AccountOverride {
            code: Some(EphemeralStorageLens::DEPLOYED_BYTECODE.clone()),
            ..Default::default()
        },
    )]);
    let lens = EphemeralStorageLens::new(address, provider);
    let call_builder = lens.extsload(slots).state(state);
    let call_builder = match block_id {
        Some(block_id) => call_builder.block(block_id),
        None => call_builder,
    };
    Ok(call_builder.call().await?._0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tests::*;
    use alloy::primitives::{address, U256};
    use futures::future::join_all;

    const POOL_ADDRESS: Address = address!("88e6A0c2dDD26FEEb64F039a2c41296FcB3f5640");

    #[tokio::test]
    async fn test_get_storage_at() {
        let provider = PROVIDER.clone();
        let slots = get_storage_at(
            POOL_ADDRESS,
            (0..10)
                .map(|i| B256::from(U256::from_limbs([i, 0, 0, 0])))
                .collect(),
            provider.clone(),
            Some(BLOCK_NUMBER),
        )
        .await
        .unwrap();
        let slots_ref = slots.as_slice();
        let provider = provider.root();
        let futures = (0..10).map(|i| async move {
            let slot = provider
                .get_storage_at(POOL_ADDRESS, U256::from_limbs([i, 0, 0, 0]))
                .block_id(BLOCK_NUMBER)
                .await
                .unwrap();
            assert_eq!(slot, U256::from_be_bytes(slots_ref[i as usize].0));
        });
        join_all(futures).await;
    }
}
